### PR TITLE
[GEOS-9534] Add percentages to sldService classify output

### PR DESCRIPTION
**Pull request by taba90**
<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [x] Committs changing the REST API, or any configuration object, should check it the REST API docs (Swagger YAML files and classic documentation) need to be updated.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by taba90** [comment]
Related Jira ticket: https://osgeo-org.atlassian.net/browse/GEOS-9534

Depends upon:

- [GEOT-6526](https://osgeo-org.atlassian.net/browse/GEOT-6526)  https://github.com/geotools/geotools/pull/2829
-[JAI-EXT-260](https://github.com/geosolutions-it/jai-ext/issues/260)  https://github.com/geosolutions-it/jai-ext/pull/261

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java [hunk 1]
```diff
@@ -370,7 +424,7 @@ public ColorMap createCustomColorMap(
             breaks[i + 1] = Converters.convert(classifier.getMax(i), Double.class);
         }
 
-        return getColorMapFromBreaks(breaks, open, continuous);
+        return getColorMapFromBreaks(breaks, open, continuous, null);
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java]: If instructed to build percentages while doing a custom color map, they should show up. It's possible to calculate them using a histogram operation.

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java [hunk 2]
```diff
@@ -418,4 +472,12 @@ private NumberRange getOperationRange(ImageWorker iw) {
             }
         }
     }
+
+    private String getPercentagesLabelPortion(double[] percentages, int i) {
+        StringBuilder sb = new StringBuilder();
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java]: Overly complicated:

```
if (percentages == null) return "";
return " (" + percentages[i] + ")";
```

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java [hunk 3]
```diff
@@ -86,7 +92,11 @@ public void setIncludeStrokeForPoints(boolean includeStrokeForPoints) {
             String functionName) {
         try {
             final Function classify =
-                    FF.function(functionName, FF.property(property), FF.literal(classNumber));
+                    FF.function(
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: Problematic with that function having two optional values at the end (see the GeoTools review)

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java [hunk 4]
```diff
@@ -315,7 +325,8 @@ public StyleFactory getStyleFactory() {
         Filter f;
         List<Rule> list = new ArrayList();
         Expression att = normalizeProperty(FF.property(property), propertyType, normalize);
-
+        PercentagesManager percMan = null;
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: Why split declaration and assignment over two lines?

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java (hunk 2 above)
Comment (aaime) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java]: By the way, the vector code states it's a percentage, e..g adds " (x%)" while this one just adds " (x)". I think it's best to have the % sign, the number per se means nothing to the ones just looking at the map without knowing how the the SLD classification service worked, but at the very least, they should be consistent.

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java [hunk 5]
```diff
@@ -472,8 +496,16 @@ private Expression normalizeProperty(
         return null;
     }
 
-    private boolean isDuplicatedClass(List<Rule> rules, Filter f) {
-        return rules.stream().anyMatch(r -> r.getFilter().equals(f));
+    private boolean isDuplicatedClass(
+            List<Rule> rules, Filter f, PercentagesManager percMan, int currentIdx) {
+        Optional<Rule> opRule = rules.stream().filter(r -> r.getFilter().equals(f)).findFirst();
+        boolean result = opRule.isPresent();
+        if (percMan != null) {
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: Two nested if can be expressed as one if with and-ed conditions

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java [hunk 6]
```diff
@@ -499,4 +531,45 @@ private Filter getNotOverlappingFilter(
...
+        boolean collapsePercentages;
+
+        public PercentagesManager(double[] percentages) {
+            this.percentages = percentages;
+            this.collapsePercentages =
+                    percentages != null && !(DoubleStream.of(percentages).sum() > 100.0);
+            this.collapsedPercentages = new ArrayList<>();
+        }
+
+        public void collapsePercentages(List<Rule> rules, Rule current, int index) {
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]:
```suggestion
        public void collapseRulePercentage(List<Rule> rules, Rule current, int index) {
```

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java [hunk 7]
```diff
@@ -499,4 +531,45 @@ private Filter getNotOverlappingFilter(
...
+            if (outputPercentages && collapsePercentages) {
+                int ruleIdx = rules.indexOf(current);
+                double toSumTo = percentages[index];
+                double toBeSummed = collapsedPercentages.get(ruleIdx);
+                collapsedPercentages.set(ruleIdx, toBeSummed + toSumTo);
+            }
+        }
+
+        public void addPercentages(int currentIndex) {
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]:
```suggestion
        public void collectRulePercentage(int currentIndex) {
```
(adding might seem a sum)

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java [hunk 8]
```diff
@@ -499,4 +531,45 @@ private Filter getNotOverlappingFilter(
...
+            if (outputPercentages) collapsedPercentages.add(percentages[currentIndex]);
+        }
+
+        public void addPercentagesToLabel(List<Rule> rules) {
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]:
```suggestion
        public void appendPercentagesToLabels(List<Rule> rules) {
```

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java [hunk 9]
```diff
@@ -499,4 +531,45 @@ private Filter getNotOverlappingFilter(
...
+            if (outputPercentages) {
+                for (int i = 0; i < rules.size(); i++) {
+                    Rule rule = rules.get(i);
+                    StringBuffer sb =
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: Java builds a StringBuffer every time you concatenate strings, avoid building it explicitly unless you're working on a loop/conditional where multiple string concatenations are split among different commands. 
For this one using:
``String title = rule.getDescription().getTitle() + " (" + collapsedPercentanges.get(i) + "%)"``
is just as efficient, but more readable.

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java [hunk 10]
```diff
@@ -2128,4 +2130,195 @@ private void checkNotOverlappingRules(Rule first, Rule second) throws IOExceptio
         assertTrue(first.getFilter().evaluate(feature));
         assertFalse(second.getFilter().evaluate(feature));
     }
+
+    @Test
+    public void testPercentagesInRulesLabelsVectors() throws Exception {
```
Comment (aaime) [inline src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: Please split the tests in one test method per check, and validate the computed percentage values too.

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java [hunk 11]
```diff
@@ -2128,4 +2130,195 @@ private void checkNotOverlappingRules(Rule first, Rule second) throws IOExceptio
...
+                                .replace("</Rules>", sldPostfix));
+        assertTrue(rulesEqual.length == 3);
+        for (Rule r : rulesEqual) {
+            Matcher rgxMatcher = rgx.matcher(r.getDescription().getTitle());
+            assertTrue(rgxMatcher.find());
+        }
+    }
+
+    @Test
+    public void testPercentagesInRulesLabelsRaster() throws Exception {
```
Comment (aaime) [inline src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: Same as above.

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java (hunk 10 above)
Comment (aaime) [reply src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: Please add a test for user provided classes as well, both raster and vector cases.
Also, these are all "happy paths", can you consider tests where things go wrong (e.g., rules that need to be merged, think of more) and add those as well?

File: doc/en/user/source/extensions/sldservice/index.rst [hunk 12]
```diff
@@ -218,6 +218,11 @@ The parameters usable to customize the ColorMap are:
      - used only for raster layers, if set to true will generate a raster pallette that interpolates linearly between classified values 
      - true|false
      -
+   * - percentages
+     - allows to obtain percentages of values in each class. For raster layers the percentages will be outputted in the label of the ColorMapEntry,
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/extensions/sldservice/index.rst]: fixed

File: doc/en/user/source/extensions/sldservice/index.rst [hunk 13]
```diff
@@ -218,6 +218,11 @@ The parameters usable to customize the ColorMap are:
      - used only for raster layers, if set to true will generate a raster pallette that interpolates linearly between classified values 
      - true|false
      -
+   * - percentages
+     - allows to obtain percentages of values in each class. For raster layers the percentages will be outputted in the label of the ColorMapEntry, 
+       while for vector layer they will  be placed in the rule title; In both case the case they will be placed at then end of the text between parenthesis.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/extensions/sldservice/index.rst]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java [hunk 14]
```diff
@@ -147,6 +153,21 @@ public ColorMap uniqueIntervalClassification(RenderedImage image, Integer maxInt
                             + maxIntervals
                             + " was requested");
         }
+        if (outputPercentages) {
+            double total = IntStream.of(bins).sum();
+            ColorMapEntry[] cEntries = colorMap.getColorMapEntries();
+            for (int i = 0; i < cEntries.length; i++) {
+                ColorMapEntry entry = cEntries[i];
+                double classMembers = (double) bins[i];
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java (hunk 1 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java (hunk 2 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java (hunk 3 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java (hunk 4 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java (hunk 5 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java (hunk 6 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java (hunk 7 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java (hunk 8 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java (hunk 9 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RulesBuilder.java]: fixed

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java (hunk 11 above)
Comment (taba90) [reply src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: fixed

File: src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java (hunk 10 above)
Comment (taba90) [reply src/extension/sldService/src/test/java/org/geoserver/sldservice/rest/ClassifierTest.java]: fixed

**Comment by taba90** [comment]
thanks for the review @aaime all feedbacks have been applied

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java [hunk 15]
```diff
@@ -418,4 +475,30 @@ private NumberRange getOperationRange(ImageWorker iw) {
             }
         }
     }
+
+    private String getPercentagesLabelPortion(double[] percentages, int i) {
+        if (percentages == null) return "";
+        else return " (" + percentages[i] + "%)";
+    }
+
+    private double[] getPercentagesForCustom(RenderedImage image, int intervals) {
```
Comment (aaime) [inline src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java]: This seems to assume an equal interval approach, but with the custom classification the client is passing in the break values instead, they can be distributed in whatever way. User provided breaks should be used, please add a test with custom values and a non uniformly distributed set of breaks.

File: doc/en/user/source/extensions/sldservice/index.rst [hunk 16]
```diff
@@ -218,6 +218,14 @@ The parameters usable to customize the ColorMap are:
...
      - true|false
      -
+   * - percentages
+     - allows to obtain percentages of values in each class. For raster layers the percentages will be included in the label of the ColorMapEntry, 
+       while for vector layer they will  be placed in the rule title; in both the cases they will be placed at then end of the text between parenthesis.
+     - true|false
+     - 
+   * - percentagesScale
+     - number of digits of percentages.
+     -
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/extensions/sldservice/index.rst]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/PercentagesRoundHandler.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+ * application directory.
+ */
+package org.geoserver.sldservice.utils.classifier;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.List;
+import java.util.stream.DoubleStream;
+
+public class PercentagesRoundHandler {
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/PercentagesRoundHandler.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/PercentagesRoundHandler.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    public PercentagesRoundHandler(Integer scale) {
+        this.scale = scale != null ? scale : 1;
+    }
+
+    public PercentagesRoundHandler() {
+        this.scale = 1;
+    }
+
+    public List<Double> roundPercentages(List<Double> percentages) {
+        if (percentages == null || percentages.size() == 0) return null;
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/PercentagesRoundHandler.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/PercentagesRoundHandler.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+        int size = percentages.size();
+        for (int i = 0; i < size; i++) {
+            double percentage = percentages.get(i);
+            double rounded = roundDouble(percentage);
+            percentages.set(i, rounded);
+        }
+        double delta = 100.0 - percentages.stream().mapToDouble(d -> d).sum();
+        if (delta > 0.0) {
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/PercentagesRoundHandler.java]: fixed

File: src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java (hunk 15 above)
Comment (taba90) [reply src/extension/sldService/src/main/java/org/geoserver/sldservice/utils/classifier/RasterSymbolizerBuilder.java]: fixed

**Comment by aaime** [comment]
So this failure:

```
[ERROR] Tests run: 85, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 29.383 s <<< FAILURE! - in org.geoserver.sldservice.rest.ClassifierTest
[ERROR] testPercentagesInRulesLabelsVectorsEqualIntervalWithOutlier(org.geoserver.sldservice.rest.ClassifierTest)  Time elapsed: 0.08 s  <<< ERROR!
org.geoserver.platform.ServiceException: expected status <200> but was <404>:Could not generate any rule, there is likely no data matching the request (layer is empty, of filtered down to no matching features/pixels)
	at org.geoserver.sldservice.rest.ClassifierTest.testPercentagesInRulesLabelsVectorsEqualIntervalWithOutlier(ClassifierTest.java:2228)
```

is due to the GeoTools fix I've just merged (and that was not yet in when this build was run)?

**Comment by taba90** [comment]
> So this failure:
> 
> ```
> [ERROR] Tests run: 85, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 29.383 s <<< FAILURE! - in org.geoserver.sldservice.rest.ClassifierTest
> [ERROR] testPercentagesInRulesLabelsVectorsEqualIntervalWithOutlier(org.geoserver.sldservice.rest.ClassifierTest)  Time elapsed: 0.08 s  <<< ERROR!
> org.geoserver.platform.ServiceException: expected status <200> but was <404>:Could not generate any rule, there is likely no data matching the request (layer is empty, of filtered down to no matching features/pixels)
> 	at org.geoserver.sldservice.rest.ClassifierTest.testPercentagesInRulesLabelsVectorsEqualIntervalWithOutlier(ClassifierTest.java:2228)
> ```
> 
> is due to the GeoTools fix I've just merged (and that was not yet in when this build was run)?

yes @aaime now the build should be succesfull

**Pull request merged**